### PR TITLE
Fix/contractaddr

### PR DESCRIPTION
**Pull request by renlulu**
## Description
This PR is to revert the breaking change for contract.call(), and also provide another way to init deployed contract if people do want to check the format of contract address.

## Review Suggestion
1. Run unit tests
2. Test base16 address:

```typescript
//  Copyright (C) 2018 Zilliqa
//
//  This file is part of Zilliqa-Javascript-Library.
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <https://www.gnu.org/licenses/>.

const { BN, Long, bytes, units } = require('@zilliqa-js/util');
const { Zilliqa } = require('@zilliqa-js/zilliqa');
const {
  toBech32Address,
  getAddressFromPrivateKey,
} = require('@zilliqa-js/crypto');

const zilliqa = new Zilliqa('https://dev-api.zilliqa.com');

// These are set by the core protocol, and may vary per-chain.
// You can manually pack the bytes according to chain id and msg version.
// For more information: https://apidocs.zilliqa.com/?shell#getnetworkid

const chainId = 333; // chainId of the developer testnet
const msgVersion = 1; // current msgVersion
const VERSION = bytes.pack(chainId, msgVersion);

// Populate the wallet with an account
const privateKey =
  'deb5c896228f8515146aa16f94a558ba14e52d8496b4b267b2d59cd9036f39a6';

zilliqa.wallet.addByPrivateKey(privateKey);

const address = getAddressFromPrivateKey(privateKey);
console.log(`My account address is: ${address}`);
console.log(`My account bech32 address is: ${toBech32Address(address)}`);

async function testBlockchain() {
  try {
    // Get Balance
    const balance = await zilliqa.blockchain.getBalance(address);
    // Get Minimum Gas Price from blockchain
    const minGasPrice = await zilliqa.blockchain.getMinimumGasPrice();

    // Account balance (See note 1)
    console.log(`Your account balance is:`);
    console.log(balance.result);
    console.log(`Current Minimum Gas Price: ${minGasPrice.result}`);
    const myGasPrice = units.toQa('1000', units.Units.Li); // Gas Price that will be used by all transactions
    console.log(`My Gas Price ${myGasPrice.toString()}`);
    const isGasSufficient = myGasPrice.gte(new BN(minGasPrice.result)); // Checks if your gas price is less than the minimum gas price
    console.log(`Is the gas price sufficient? ${isGasSufficient}`);

    const addr = '0x669725960f9F3A77Ad8ce9214DA91ED4661322d9'.toLocaleLowerCase();
    console.log(addr);
    const deployedContract = zilliqa.contracts.at(
      addr
    );

    // Create a new timebased message and call setHello
    // Also notice here we have a default function parameter named toDs as mentioned above.
    // For calling a smart contract, any transaction can be processed in the DS but not every transaction can be processed in the shards.
    // For those transactions are involved in chain call, the value of toDs should always be true.
    // If a transaction of contract invocation is sent to a shard and if the shard is not allowed to process it, then the transaction will be dropped.
    const newMsg = 'Hello, the time is ' + Date.now();
    console.log('Calling setHello transition with msg: ' + newMsg);
    const callTx = await deployedContract.callWithoutConfirm(
      'setHello',
      [
        {
          vname: 'msg',
          type: 'String',
          value: newMsg,
        },
      ],
      {
        // amount, gasPrice and gasLimit must be explicitly provided
        version: VERSION,
        amount: new BN(0),
        gasPrice: myGasPrice,
        gasLimit: Long.fromNumber(8000),
      },
      false,
    );

    // check the pending status
    const pendingStatus = await zilliqa.blockchain.getPendingTxn(callTx.id);
    console.log(`Pending status is: `);
    console.log(pendingStatus.result);

    // process confirm
    console.log(`The transaction id is:`, callTx.id);
    console.log(`Waiting transaction be confirmed`);
    const confirmedTxn = await callTx.confirm(callTx.id);

    console.log(`The transaction status is:`);
    console.log(confirmedTxn.receipt);
    if (confirmedTxn.receipt.success === true) {
      console.log(`Contract address is: ${deployedContract.address}`);
    }
  } catch (err) {
    console.log(err);
  }
}

testBlockchain();
```

3. Test checksum address

```typescript
//  Copyright (C) 2018 Zilliqa
//
//  This file is part of Zilliqa-Javascript-Library.
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <https://www.gnu.org/licenses/>.

const { BN, Long, bytes, units } = require('@zilliqa-js/util');
const { Zilliqa } = require('@zilliqa-js/zilliqa');
const {
  toBech32Address,
  getAddressFromPrivateKey,
} = require('@zilliqa-js/crypto');

const zilliqa = new Zilliqa('https://dev-api.zilliqa.com');

// These are set by the core protocol, and may vary per-chain.
// You can manually pack the bytes according to chain id and msg version.
// For more information: https://apidocs.zilliqa.com/?shell#getnetworkid

const chainId = 333; // chainId of the developer testnet
const msgVersion = 1; // current msgVersion
const VERSION = bytes.pack(chainId, msgVersion);

// Populate the wallet with an account
const privateKey =
  'deb5c896228f8515146aa16f94a558ba14e52d8496b4b267b2d59cd9036f39a6';

zilliqa.wallet.addByPrivateKey(privateKey);

const address = getAddressFromPrivateKey(privateKey);
console.log(`My account address is: ${address}`);
console.log(`My account bech32 address is: ${toBech32Address(address)}`);

async function testBlockchain() {
  try {
    // Get Balance
    const balance = await zilliqa.blockchain.getBalance(address);
    // Get Minimum Gas Price from blockchain
    const minGasPrice = await zilliqa.blockchain.getMinimumGasPrice();

    // Account balance (See note 1)
    console.log(`Your account balance is:`);
    console.log(balance.result);
    console.log(`Current Minimum Gas Price: ${minGasPrice.result}`);
    const myGasPrice = units.toQa('1000', units.Units.Li); // Gas Price that will be used by all transactions
    console.log(`My Gas Price ${myGasPrice.toString()}`);
    const isGasSufficient = myGasPrice.gte(new BN(minGasPrice.result)); // Checks if your gas price is less than the minimum gas price
    console.log(`Is the gas price sufficient? ${isGasSufficient}`);

    const addr = '0x669725960f9F3A77Ad8ce9214DA91ED4661322d9';
    console.log(addr);
    const deployedContract = zilliqa.contracts.at(
      addr
    );

    // Create a new timebased message and call setHello
    // Also notice here we have a default function parameter named toDs as mentioned above.
    // For calling a smart contract, any transaction can be processed in the DS but not every transaction can be processed in the shards.
    // For those transactions are involved in chain call, the value of toDs should always be true.
    // If a transaction of contract invocation is sent to a shard and if the shard is not allowed to process it, then the transaction will be dropped.
    const newMsg = 'Hello, the time is ' + Date.now();
    console.log('Calling setHello transition with msg: ' + newMsg);
    const callTx = await deployedContract.callWithoutConfirm(
      'setHello',
      [
        {
          vname: 'msg',
          type: 'String',
          value: newMsg,
        },
      ],
      {
        // amount, gasPrice and gasLimit must be explicitly provided
        version: VERSION,
        amount: new BN(0),
        gasPrice: myGasPrice,
        gasLimit: Long.fromNumber(8000),
      },
      false,
    );

    // check the pending status
    const pendingStatus = await zilliqa.blockchain.getPendingTxn(callTx.id);
    console.log(`Pending status is: `);
    console.log(pendingStatus.result);

    // process confirm
    console.log(`The transaction id is:`, callTx.id);
    console.log(`Waiting transaction be confirmed`);
    const confirmedTxn = await callTx.confirm(callTx.id);

    console.log(`The transaction status is:`);
    console.log(confirmedTxn.receipt);
    if (confirmedTxn.receipt.success === true) {
      console.log(`Contract address is: ${deployedContract.address}`);
    }
  } catch (err) {
    console.log(err);
  }
}

testBlockchain();

```

4. Test bech32 address

```typescript
//  Copyright (C) 2018 Zilliqa
//
//  This file is part of Zilliqa-Javascript-Library.
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <https://www.gnu.org/licenses/>.

const { BN, Long, bytes, units } = require('@zilliqa-js/util');
const { Zilliqa } = require('@zilliqa-js/zilliqa');
const {
  toBech32Address,
  getAddressFromPrivateKey,
} = require('@zilliqa-js/crypto');

const zilliqa = new Zilliqa('https://dev-api.zilliqa.com');

// These are set by the core protocol, and may vary per-chain.
// You can manually pack the bytes according to chain id and msg version.
// For more information: https://apidocs.zilliqa.com/?shell#getnetworkid

const chainId = 333; // chainId of the developer testnet
const msgVersion = 1; // current msgVersion
const VERSION = bytes.pack(chainId, msgVersion);

// Populate the wallet with an account
const privateKey =
  'deb5c896228f8515146aa16f94a558ba14e52d8496b4b267b2d59cd9036f39a6';

zilliqa.wallet.addByPrivateKey(privateKey);

const address = getAddressFromPrivateKey(privateKey);
console.log(`My account address is: ${address}`);
console.log(`My account bech32 address is: ${toBech32Address(address)}`);

async function testBlockchain() {
  try {
    // Get Balance
    const balance = await zilliqa.blockchain.getBalance(address);
    // Get Minimum Gas Price from blockchain
    const minGasPrice = await zilliqa.blockchain.getMinimumGasPrice();

    // Account balance (See note 1)
    console.log(`Your account balance is:`);
    console.log(balance.result);
    console.log(`Current Minimum Gas Price: ${minGasPrice.result}`);
    const myGasPrice = units.toQa('1000', units.Units.Li); // Gas Price that will be used by all transactions
    console.log(`My Gas Price ${myGasPrice.toString()}`);
    const isGasSufficient = myGasPrice.gte(new BN(minGasPrice.result)); // Checks if your gas price is less than the minimum gas price
    console.log(`Is the gas price sufficient? ${isGasSufficient}`);

    const addr = 'zil1v6tjt9s0nua80tvvays5m2g763npxgkez0gnnq';
    console.log(addr);
    const deployedContract = zilliqa.contracts.at(
      addr
    );

    // Create a new timebased message and call setHello
    // Also notice here we have a default function parameter named toDs as mentioned above.
    // For calling a smart contract, any transaction can be processed in the DS but not every transaction can be processed in the shards.
    // For those transactions are involved in chain call, the value of toDs should always be true.
    // If a transaction of contract invocation is sent to a shard and if the shard is not allowed to process it, then the transaction will be dropped.
    const newMsg = 'Hello, the time is ' + Date.now();
    console.log('Calling setHello transition with msg: ' + newMsg);
    const callTx = await deployedContract.callWithoutConfirm(
      'setHello',
      [
        {
          vname: 'msg',
          type: 'String',
          value: newMsg,
        },
      ],
      {
        // amount, gasPrice and gasLimit must be explicitly provided
        version: VERSION,
        amount: new BN(0),
        gasPrice: myGasPrice,
        gasLimit: Long.fromNumber(8000),
      },
      false,
    );

    // check the pending status
    const pendingStatus = await zilliqa.blockchain.getPendingTxn(callTx.id);
    console.log(`Pending status is: `);
    console.log(pendingStatus.result);

    // process confirm
    console.log(`The transaction id is:`, callTx.id);
    console.log(`Waiting transaction be confirmed`);
    const confirmedTxn = await callTx.confirm(callTx.id);

    console.log(`The transaction status is:`);
    console.log(confirmedTxn.receipt);
    if (confirmedTxn.receipt.success === true) {
      console.log(`Contract address is: ${deployedContract.address}`);
    }
  } catch (err) {
    console.log(err);
  }
}

testBlockchain();

```

5. Force checking address format by using `contract.atBech32`

```typescript
//  Copyright (C) 2018 Zilliqa
//
//  This file is part of Zilliqa-Javascript-Library.
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <https://www.gnu.org/licenses/>.

const { BN, Long, bytes, units } = require('@zilliqa-js/util');
const { Zilliqa } = require('@zilliqa-js/zilliqa');
const {
  toBech32Address,
  getAddressFromPrivateKey,
} = require('@zilliqa-js/crypto');

const zilliqa = new Zilliqa('https://dev-api.zilliqa.com');

// These are set by the core protocol, and may vary per-chain.
// You can manually pack the bytes according to chain id and msg version.
// For more information: https://apidocs.zilliqa.com/?shell#getnetworkid

const chainId = 333; // chainId of the developer testnet
const msgVersion = 1; // current msgVersion
const VERSION = bytes.pack(chainId, msgVersion);

// Populate the wallet with an account
const privateKey =
  'deb5c896228f8515146aa16f94a558ba14e52d8496b4b267b2d59cd9036f39a6';

zilliqa.wallet.addByPrivateKey(privateKey);

const address = getAddressFromPrivateKey(privateKey);
console.log(`My account address is: ${address}`);
console.log(`My account bech32 address is: ${toBech32Address(address)}`);

async function testBlockchain() {
  try {
    // Get Balance
    const balance = await zilliqa.blockchain.getBalance(address);
    // Get Minimum Gas Price from blockchain
    const minGasPrice = await zilliqa.blockchain.getMinimumGasPrice();

    // Account balance (See note 1)
    console.log(`Your account balance is:`);
    console.log(balance.result);
    console.log(`Current Minimum Gas Price: ${minGasPrice.result}`);
    const myGasPrice = units.toQa('1000', units.Units.Li); // Gas Price that will be used by all transactions
    console.log(`My Gas Price ${myGasPrice.toString()}`);
    const isGasSufficient = myGasPrice.gte(new BN(minGasPrice.result)); // Checks if your gas price is less than the minimum gas price
    console.log(`Is the gas price sufficient? ${isGasSufficient}`);

    const addr = 'zil1v6tjt9s0nua80tvvays5m2g763npxgkez0gnnq';
    console.log(addr);
    const deployedContract = zilliqa.contracts.atBech32(
      addr
    );

    // Create a new timebased message and call setHello
    // Also notice here we have a default function parameter named toDs as mentioned above.
    // For calling a smart contract, any transaction can be processed in the DS but not every transaction can be processed in the shards.
    // For those transactions are involved in chain call, the value of toDs should always be true.
    // If a transaction of contract invocation is sent to a shard and if the shard is not allowed to process it, then the transaction will be dropped.
    const newMsg = 'Hello, the time is ' + Date.now();
    console.log('Calling setHello transition with msg: ' + newMsg);
    const callTx = await deployedContract.callWithoutConfirm(
      'setHello',
      [
        {
          vname: 'msg',
          type: 'String',
          value: newMsg,
        },
      ],
      {
        // amount, gasPrice and gasLimit must be explicitly provided
        version: VERSION,
        amount: new BN(0),
        gasPrice: myGasPrice,
        gasLimit: Long.fromNumber(8000),
      },
      false,
    );

    // check the pending status
    const pendingStatus = await zilliqa.blockchain.getPendingTxn(callTx.id);
    console.log(`Pending status is: `);
    console.log(pendingStatus.result);

    // process confirm
    console.log(`The transaction id is:`, callTx.id);
    console.log(`Waiting transaction be confirmed`);
    const confirmedTxn = await callTx.confirm(callTx.id);

    console.log(`The transaction status is:`);
    console.log(confirmedTxn.receipt);
    if (confirmedTxn.receipt.success === true) {
      console.log(`Contract address is: ${deployedContract.address}`);
    }
  } catch (err) {
    console.log(err);
  }
}

testBlockchain();

```
### Implementation

- [x] **ready for review**

### CODE REVIEW

File: packages/zilliqa-js-contract/src/contract.ts [hunk 1]
```diff
@@ -147,6 +159,8 @@ export class Contract {
       { ...tx.txParams, priority: tx.toDS },
     );
 
+    console.log(response);
```
Comment (teye) [inline packages/zilliqa-js-contract/src/contract.ts:162]: remove this console.log?

Comment (renlulu) [reply packages/zilliqa-js-contract/src/contract.ts:162]: Oh! OK!

**Comment by bb111189** [comment]
hi @bogdan 

Thanks for your feedback. We are planning to update the SDK with this fix. I think it will resolve the breaking change introduced in the previous version.

**Comment by bogdan** [comment]
@bb111189 @renlulu I also had problems with `zilliqa.transactions.new({toAddr})`. I didn't accept bech32 in version 0.11.0 too. Are you sure this PR fixes it too?

**Comment by renlulu** [comment]
> @bb111189 @renlulu I also had problems with `zilliqa.transactions.new({toAddr})`. I didn't accept bech32 in version 0.11.0 too. Are you sure this PR fixes it too?

I don't think that one need to be `fixed`, because initially we found some people would transfer their zils to their ethereum address by mistake, that's they we decided to make the change later. If you do not want to convert your address to bech32 or checksum format, I can provide another function something like `newBase16()`, but I cannot change `new()` anymore. It already be used by several exchanges for security purposes.

**Comment by bogdan** [comment]
@renlulu I don't want to restrict how your software should work, but I really don't appreciate breaking changes and want to make sure Zilliqa Team shares this common vision of the community described in https://semver.org/

So based on that definition it needs to be `fixed`. The security argument is strong and I understand that. But you could do it gracefully by introduce something like `new Zilliqa({enableSecureToAddress: true ,.... })` or a warning like `{toAddr} will only accept bech32 address from zilliiqa.js version 1.0.0` in order to avoid breaking change.

**Comment by renlulu** [comment]
> @renlulu I don't want to restrict how your software should work, but I really don't appreciate breaking changes and want to make sure Zilliqa Team shares this common vision of the community described in https://semver.org/
> 
> So based on that definition it needs to be `fixed`. The security argument is strong and I understand that. But you could do it gracefully by introduce something like `new Zilliqa({enableSecureToAddress: true ,.... })` or a warning like `{toAddr} will only accept bech32 address from zilliiqa.js version 1.0.0` in order to avoid breaking change.

Alright, we didn't upgrade our `Major` version. I think we can follow your suggestion by introducing `enableSecureToAdress`.

**Comment by bogdan** [comment]
Thanks for that. 
Just in case: if you would just upgrade the `Major` version - it wouldn't be good enough.

The adequate solution would be to introduce a deprecation warning in `0.x` series like: `{toAddr} will only accept bech32 address from zilliiqa.js version 1.0.0` and then handle a breaking change in `1.x.x` series and support it with a changelog or release notes for 1.0.

Another adequate solution would be to introduce `enableSecureToAddress` with a warning that it will not be an option in `1.x` and will be enabled for everyone all the time.

This is very important for us because it seems like we relay on the library like no one else and we require certain level of stability and graceful introduction of changes.

**Comment by renlulu** [comment]
> Thanks for that.
> Just in case: if you would just upgrade the `Major` version - it wouldn't be good enough.
> 
> The adequate solution would be to introduce a deprecation warning in `0.x` series like: `{toAddr} will only accept bech32 address from zilliiqa.js version 1.0.0` and then handle a breaking change in `1.x.x` series and support it with a changelog or release notes for 1.0.
> 
> Another adequate solution would be to introduce `enableSecureToAddress` with a warning that it will not be an option in `1.x` and will be enabled for everyone all the time.
> 
> This is very important for us because it seems like we relay on the library like no one else and we require certain level of stability and graceful introduction of changes.

Thanks for you advice, it is really solid suggestion that we should keep in mind. For `enableSecureToAddress`, my idea is just set it a default value, which is true, meaning that people would need set it to `false` explicitly if they `do` want to pass a non bech32 or checksum address. I think for your case, you will need to do that too. Any comment?

**Comment by renlulu** [comment]
> > Thanks for that.
> > Just in case: if you would just upgrade the `Major` version - it wouldn't be good enough.
> > The adequate solution would be to introduce a deprecation warning in `0.x` series like: `{toAddr} will only accept bech32 address from zilliiqa.js version 1.0.0` and then handle a breaking change in `1.x.x` series and support it with a changelog or release notes for 1.0.
> > Another adequate solution would be to introduce `enableSecureToAddress` with a warning that it will not be an option in `1.x` and will be enabled for everyone all the time.
> > This is very important for us because it seems like we relay on the library like no one else and we require certain level of stability and graceful introduction of changes.
> 
> Thanks for you advice, it is really solid suggestion that we should keep in mind. For `enableSecureToAddress`, my idea is just set it a default value, which is true, meaning that people would need set it to `false` explicitly if they `do` want to pass a non bech32 or checksum address. I think for your case, you will need to do that too. Any comment?

@bogdan Pushed, let me know if you still have any concern, thanks!

**Comment by bogdan** [comment]
This one looks backward compatible to me.

**Pull request merged**